src/draw/shapes.typ

```diff
@@ -20,6 +20,45 @@
 #import "styling.typ": *
 #import "grouping.typ": *
 
+
+// Apply path decoration function fn to drawables
+// and return a single drawable.
+#let _apply-decoration-fn(ctx, style, drawables, fn) = {
+  if fn != none {
+    ctx.transform = matrix.ident()
+    let (drawables: drawables, ..) = process.many(ctx, (fn)((ctx => {
+      return (ctx: ctx, drawables: if type(drawables) != array { (drawables,) } else { drawables })
+    },), ..style, decoration: none))
+
+    assert.eq(drawables.len(), 1,
+      message: "Path decoration function must return a single element")
+    return drawables.first()
+  }
+  return drawables
+}
+
+// Apply path decoration to drawables
+#let _apply-decoration(ctx, style, drawables) = {
+  let deco = style.at("decoration", default: none)
+
+  if deco != none {
+    import "/src/lib/decorations.typ": wave, zigzag, coil
+    assert(type(deco) in (str, function),
+      message: "Decoration must be of type string or function")
+
+    let fn = if type(deco) == str {
+      ctx.decorations.at(deco)
+    } else if type(deco) == function {
+      deco
+    }
+
+    assert(fn != none,
+      message: "Decoration function is none")
+    return _apply-decoration-fn(ctx, style, drawables, fn)
+  }
+  return drawables
+}
```

Comment:
Move these functions to `lib/decorations.typ` even if they should only be used with shape elements they can still be used to apply decorations to other elements.

Reply:
Yes. Do we want to have some kind of function to apply modifiers to a path post construction? Some generic function that accepts a modifier + path. The existing deco functions would then also use that. We should move the implementation out of lib, as functions from lib should not be used by core cetz.

A modifier function should be specified as a function taking a context, styles and a single drawable and outputs a single drawable.

Reply:
We should discuss this further in #467